onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp

```diff
@@ -0,0 +1,807 @@
+        )
+    {
+        Accumulators[Row][0] = MlasZeroFloat32x4();
+        Accumulators[Row][1] = MlasZeroFloat32x4();
+        Accumulators[Row][2] = MlasZeroFloat32x4();
+        Accumulators[Row][3] = MlasZeroFloat32x4();
+    }
+};
+
+struct MlasSgemmLoadAElements
+{
+    template<size_t RowCount, size_t Row>
+    MLAS_FORCEINLINE
+    static
+    void
+    Iteration(
+        MLAS_FLOAT32X4 AElements[RowCount],
+        const float* A,
+        size_t lda
+        )
+    {
+        AElements[Row] = MlasLoadFloat32x4(A + Row * lda);
+    }
+};
+struct MlasSgemmBroadcastAElementsMMA
+{
+    template<size_t RowCount, size_t Row>
+    MLAS_FORCEINLINE
+    static
+    void
+    Iteration(
+        MLAS_FLOAT32X4 ABroadcast[RowCount],
+        const float* A,
+        size_t lda
+        )
+    {
+        ABroadcast[0][Row] = A [Row * lda];
+    }
+};
+struct MlasSgemmBroadcastAElements
+{
+    template<size_t RowCount, size_t Row>
+    MLAS_FORCEINLINE
+    static
+    void
+    Iteration(
+        MLAS_FLOAT32X4 ABroadcast[RowCount],
+        const float* A,
+        size_t lda
+        )
+    {
+        ABroadcast[Row] = MlasBroadcastFloat32x4(A + Row * lda);
+    }
+};
+
+template<unsigned Lane>
+struct MlasSgemmSplatAElements
+{
+    template<size_t RowCount, size_t Row>
+    MLAS_FORCEINLINE
+    static
+    void
+    Iteration(
+        MLAS_FLOAT32X4 AElements[RowCount],
+        MLAS_FLOAT32X4 ABroadcast[RowCount]
+        )
+    {
+        ABroadcast[Row] = vec_splat(AElements[Row], Lane);
+    }
+};
+
+struct MlasSgemmMultiplyAddRow
+{
+    template<size_t RowCount, size_t Row>
+    MLAS_FORCEINLINE
+    static
+    void
+    Iteration(
+        MLAS_FLOAT32X4 Accumulators[RowCount][4],
+        MLAS_FLOAT32X4 ABroadcast[RowCount],
+        MLAS_FLOAT32X4 BElements[4]
+        )
+    {
+        Accumulators[Row][0] = MlasMultiplyAddFloat32x4(ABroadcast[Row], BElements[0], Accumulators[Row][0]);
+        Accumulators[Row][1] = MlasMultiplyAddFloat32x4(ABroadcast[Row], BElements[1], Accumulators[Row][1]);
+        Accumulators[Row][2] = MlasMultiplyAddFloat32x4(ABroadcast[Row], BElements[2], Accumulators[Row][2]);
+        Accumulators[Row][3] = MlasMultiplyAddFloat32x4(ABroadcast[Row], BElements[3], Accumulators[Row][3]);
+    }
+};
+
+template<size_t RowCount>
+MLAS_FORCEINLINE
+void
+MlasSgemmComputeBlock(
+    MLAS_FLOAT32X4 Accumulators[RowCount][4],
+    MLAS_FLOAT32X4 ABroadcast[RowCount],
+    const float* B
+    )
+{
+    MLAS_FLOAT32X4 BElements[4];
+
+    BElements[0] = MlasLoadFloat32x4(B);
+    BElements[1] = MlasLoadFloat32x4(B + 4);
+    BElements[2] = MlasLoadFloat32x4(B + 8);
+    BElements[3] = MlasLoadFloat32x4(B + 12);
+
+    MlasLoopUnroll<RowCount, MlasSgemmMultiplyAddRow>()(Accumulators, ABroadcast, BElements);
+}
+template<size_t RowCount>
+MLAS_FORCEINLINE
+void
+MlasSgemmComputeAElements(
+    MLAS_FLOAT32X4 AElements[RowCount],
+    MLAS_FLOAT32X4 ABroadcast[RowCount]
```

Comment:
>RowCount [](http://example.com/codeflow?start=30&length=8)

are you assuming RowCount is 4 here? If I understand correctly, AElements is transposed to ABroadcast

Reply:
Yes RowCount is always 4 for MlasSgemmComputeAElements.